core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -262,6 +268,159 @@ public ShardIterator activeInitializingShardsIt(int seed) {
+                final String nodeId = entry.getKey();
+                final Optional<ResponseCollectorService.ComputedNodeStats> maybeStats = entry.getValue();
+                if (nodeId.equals(minNodeId) == false && maybeStats.isPresent()) {
+                    final ResponseCollectorService.ComputedNodeStats stats = maybeStats.get();
+                    final int updatedQueue = (minStats.queueSize + stats.queueSize) / 2;
+                    final long updatedResponse = (long) (minStats.responseTime + stats.responseTime) / 2;
+                    final long updatedService = (long) (minStats.serviceTime + stats.serviceTime) / 2;
+                    collector.addNodeStatistics(nodeId, updatedQueue, updatedResponse, updatedService);
+                }
+            }
+        }
+    }
+
+    private static List<ShardRouting> rank(List<ShardRouting> shards, final ResponseCollectorService collector,
+                                           final Map<String, Long> nodeSearchCounts) {
+        if (collector == null || nodeSearchCounts == null || shards.size() <= 1) {
+            return shards;
+        }
+
+        // Retrieve which nodes we can potentially send the query to
+        final Set<String> nodeIds = getAllNodeIds(shards);
+        final int nodeCount = nodeIds.size();
+
+        final Map<String, Optional<ResponseCollectorService.ComputedNodeStats>> nodeStats = getNodeStats(nodeIds, collector);
+
+        // Retrieve all the nodes the shards exist on
+        final Map<String, Double> nodeRanks = rankNodes(nodeStats, nodeSearchCounts);
+
+        String minNode = null;
+        ResponseCollectorService.ComputedNodeStats minStats = null;
+        // calculate the "winning" node and its stats (for adjusting other nodes later)
+        for (Map.Entry<String, Optional<ResponseCollectorService.ComputedNodeStats>> entry : nodeStats.entrySet()) {
+            final String nodeId = entry.getKey();
+            final Optional<ResponseCollectorService.ComputedNodeStats> maybeNodeStats = entry.getValue();
+            if (maybeNodeStats.isPresent()) {
+                ResponseCollectorService.ComputedNodeStats stats = maybeNodeStats.get();
+                double rank = stats.rank(nodeSearchCounts.getOrDefault(nodeId, 1L));
+                if (minStats == null || rank < minStats.rank(nodeSearchCounts.getOrDefault(minStats.nodeId, 1L))) {
+                    minStats = stats;
+                    minNode = nodeId;
+                }
+            }
+        }
+
+        // sort all shards based on the shard rank
+        ArrayList<ShardRouting> sortedShards = new ArrayList<>(shards);
+        Collections.sort(sortedShards, new NodeRankComparator(nodeRanks));
+
+        // adjust the non-winner nodes' stats so they will get a chance to receive queries
+        adjustStats(collector, nodeStats, minNode, minStats);
+        if (minNode != null) {
```

Comment:
If there are ties, then minNode might not be the node that we first try to send the shard request to. Should we update stats of the node id of the first shard of `sortedShards` instead?

Reply:
I did make this change, to update the first stats of `sortedShards`, it also turns out to be more efficient because I don't have to loop through everything twice!